zebrad/src/components/mempool/storage/eviction_list.rs

```diff
@@ -0,0 +1,140 @@
+//! [`EvictionList`] represents the transaction eviction list with
+//! efficient operations.
+use std::{
+    collections::{HashMap, VecDeque},
+    time::{Duration, Instant},
+};
+
+use zebra_chain::transaction;
+
+/// An eviction list that allows Zebra to efficiently add entries, get entries,
+/// and remove older entries in the order they were inserted.
+pub struct EvictionList {
+    // Maps each TXID in the list to the most recent instant they were added.
+    unique_entries: HashMap<transaction::Hash, Instant>,
+    // The same as `unique_entries` but in the order they were inserted.
+    ordered_entries: VecDeque<transaction::Hash>,
+    // The maximum size of `unique_entries`.
+    max_size: usize,
+    /// The mempool transaction eviction age limit.
+    /// Same as [`Config::eviction_memory_time`].
+    eviction_memory_time: Duration,
+}
+
+impl EvictionList {
+    /// Create a new [`EvictionList`] with the given maximum size and
+    /// eviction time.
+    pub fn new(max_size: usize, eviction_memory_time: Duration) -> Self {
+        Self {
+            unique_entries: Default::default(),
+            ordered_entries: Default::default(),
+            max_size,
+            eviction_memory_time,
+        }
+    }
+
+    /// Inserts a TXID in the list, keeping track of the time it was inserted.
+    ///
+    /// All entries older than [`EvictionList::eviction_memory_time`] will be removed.
+    ///
+    /// # Panics
+    ///
+    /// If the TXID is already in the list.
+    ///
+    pub fn insert(&mut self, key: transaction::Hash) {
+        // From https://zips.z.cash/zip-0401#specification:
+        // > Nodes SHOULD remove transactions from RecentlyEvicted that were evicted more than
+        // > mempoolevictionmemoryminutes minutes ago. This MAY be done periodically,
+        // > and/or just before RecentlyEvicted is accessed when receiving a transaction.
+        self.prune_old();
+        // > Add the txid and the current time to RecentlyEvicted, dropping the oldest entry
+        // > in RecentlyEvicted if necessary to keep it to at most eviction_memory_entries entries.
+        if self.len() >= self.max_size {
+            self.pop_front();
+        }
+        let value = Instant::now();
+        let old_value = self.unique_entries.insert(key, value);
+        // It should be impossible for an already-evicted transaction to be evicted
+        // again since transactions are not added to the mempool if they are evicted,
+        // and the mempool doesn't allow inserting two transactions with the same
+        // hash (they would conflict).
+        assert!(
+            old_value.is_none(),
```

Comment:
Diagnostics Nitpick: The `assert_eq!` panic message includes the actual value, but `assert!` doesn't.

```suggestion
        assert_eq!(
            old_value,
            None,
```